File: ARKOV/pathfinding/breadthFirstSearch.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <set>
#include <map>
#include <algorithm>
#include <thread>
#include <chrono>

int curx = 0;
int cury = 0;
size_t idx = 0;

std::vector<std::tuple<int, int>> que = {std::make_tuple(curx, cury)};
std::set<std::tuple<int, int>> queued = {std::make_tuple(curx, cury)};

// Format is { (target_x, target_y) : (source_x, source_y) }
// Using (-1, -1) to represent 'None'
std::map<std::tuple<int, int>, std::tuple<int, int>> came_from = {
    {std::make_tuple(curx, cury), std::make_tuple(-1, -1)}
};

/*
0 = not checked
1 = checked
2 = obstacle
3 = endpoint
4 = path 
*/

int map[10][10] = {
    {1,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {2,2,2,2,2,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,0,0,0,0,3}
};

void check_neihbors() {
    // check right
    if (curx + 1 <= 9 && map[curx + 1][cury] != 2 && map[curx + 1][cury] != 1 && queued.find(std::make_tuple(curx + 1, cury)) == queued.end()) {
        que.push_back(std::make_tuple(curx + 1, cury));
        queued.insert(std::make_tuple(curx + 1, cury));
        came_from[std::make_tuple(curx + 1, cury)] = std::make_tuple(curx, cury);
    }
    // check down
    if (cury + 1 <= 9 && map[curx][cury + 1] != 2 && map[curx][cury + 1] != 1 && queued.find(std::make_tuple(curx, cury + 1)) == queued.end()) {
        que.push_back(std::make_tuple(curx, cury + 1));
        queued.insert(std::make_tuple(curx, cury + 1));
        came_from[std::make_tuple(curx, cury + 1)] = std::make_tuple(curx, cury);
    }
    // check left
    if (curx - 1 >= 0 && map[curx - 1][cury] != 2 && map[curx - 1][cury] != 1 && queued.find(std::make_tuple(curx - 1, cury)) == queued.end()) {
        que.push_back(std::make_tuple(curx - 1, cury));
        queued.insert(std::make_tuple(curx - 1, cury));
        came_from[std::make_tuple(curx - 1, cury)] = std::make_tuple(curx, cury);
    }
    // check up
    if (cury - 1 >= 0 && map[curx][cury - 1] != 2 && map[curx][cury - 1] != 1 && queued.find(std::make_tuple(curx, cury - 1)) == queued.end()) {
        que.push_back(std::make_tuple(curx, cury - 1));
        queued.insert(std::make_tuple(curx, cury - 1));
        came_from[std::make_tuple(curx, cury - 1)] = std::make_tuple(curx, cury);
    }
}

int main() {
    bool target_found = false;

    while (map[curx][cury] != 3) {
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                std::cout << map[i][j] << " ";
            }
            std::cout << "\n";
        }
        std::cout << "\n";

        // std::this_thread::sleep_for(std::chrono::milliseconds(100)); 

        if (map[curx][cury] == 0) {
            map[curx][cury] = 1;
        }

        check_neihbors();

        idx++;

        if (idx >= que.size()) {
            std::cout << "Queue empty: No path found!\n";
            break;
        }

        curx = std::get<0>(que[idx]);
        cury = std::get<1>(que[idx]);

        if (map[curx][cury] == 3) {
            target_found = true;
        }
    }
    
    // PATH RECONSTRUCTION 
    if (target_found) {
        std::cout << "Target Found! Reconstructing path...\n\n";
        std::vector<std::tuple<int, int>> path;
        std::tuple<int, int> current = std::make_tuple(curx, cury);
        std::tuple<int, int> start_marker = std::make_tuple(-1, -1);

        // Loop backwards through the dictionary until we hit (-1, -1) (the start)
        while (current != start_marker) {
            path.push_back(current);
            current = came_from[current];
        }

        // Reverse the list so it goes from Start -> End
        std::reverse(path.begin(), path.end());

        std::cout << "Final Path Coordinates:\n";
        for (auto step : path) {
            map[std::get<0>(step)][std::get<1>(step)] = 4;
            std::cout << "(" << std::get<0>(step) << ", " << std::get<1>(step) << ")\n";
        }
        
        std::cout << "\nFinal Map:\n";
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                std::cout << map[i][j] << " ";
            }
            std::cout << "\n";
        }
    }

    return 0;
}

File: ARKOV/pathfinding/breadthFirstSearch.py
import numpy as np
import time

curx = 0
cury = 0

index = 0

xsize = 10
ysize = 10

que = [(curx,cury)]
map = np.zeros((xsize,ysize))

queued = set([(curx, cury)])

# Format is { (target_x, target_y) : (source_x, source_y) }
came_from = {(curx, cury): None}

'''
0 = not checked
1 = checked
2 = obstacle
3 = endpoint
4 = path 
'''

def make_obstacles():
    map[0][0] = 1 # place start
    map[9][9] = 3 # place end

    map[4][0] = 2 # obstacles 
    map[4][1] = 2
    map[4][2] = 2
    map[4][3] = 2
    map[4][4] = 2

def check_neihbors():
    global curx,cury,xsize,ysize
    
    # Check right
    if curx+1 < xsize and map[curx+1][cury] not in (1, 2) and (curx+1, cury) not in queued:
        que.append((curx+1,cury))
        queued.add((curx+1,cury))
        came_from[(curx+1, cury)] = (curx, cury) # Record where we came from
        
    # Check down
    if cury+1 < ysize and map[curx][cury+1] not in (1, 2) and (curx, cury+1) not in queued:
        que.append((curx,cury+1))
        queued.add((curx, cury+1))
        came_from[(curx, cury+1)] = (curx, cury)
        
    # Check left 
    if curx-1 >= 0 and map[curx-1][cury] not in (1, 2) and (curx-1, cury) not in queued:
        que.append((curx-1,cury))
        queued.add((curx-1, cury))
        came_from[(curx-1, cury)] = (curx, cury)
        
    # Check up 
    if cury-1 >= 0 and map[curx][cury-1] not in (1, 2) and (curx, cury-1) not in queued:
        que.append((curx,cury-1))
        queued.add((curx, cury-1))
        came_from[(curx, cury-1)] = (curx, cury)

def main():
    global curx,cury,index
    
    target_found = False
    
    while map[curx][cury] != 3:
        # Print map for viewing
        print(map)
        print("\n")
        
        #time.sleep(0.25) # Artificial Latency for viewing updates

        # update map when traversed
        if map[curx][cury] == 0:
             map[curx][cury] = 1

        # add neighbors to que
        check_neihbors()
        
        index += 1
        
        if index >= len(que):
            print("Queue empty: No path found!")
            break
            
        curVals = que[index]
        curx = curVals[0]
        cury = curVals[1]
        
        if map[curx][cury] == 3:
            target_found = True

    # --- NEW: PATH RECONSTRUCTION ---
    if target_found:
        print("Target Found! Reconstructing path...\n")
        path = []
        current = (curx, cury) # Start backtracking from the end
        
        # Loop backwards through our dictionary until we hit 'None' (the start)
        while current is not None:
            path.append(current)
            current = came_from[current]
            
        # Reverse the list so it goes from Start -> End
        path.reverse()
        
        print("Final Path Coordinates:")
        for step in path:
            map[step[0]][step[1]] = 4
            print(step)
        print(map)

if __name__ == "__main__":
    make_obstacles()
    main()